state/src/state.rs

```diff
@@ -0,0 +1,719 @@
+            .or(Err(Error::TrieError))?;
+        match trie.get(&address) {
+            Ok(Some(rlp)) => {
+                let state_object = StateObject::from_rlp(&rlp)?;
+                self.insert_cache(
+                    address,
+                    StateObjectEntry::new_clean(Some(state_object.clone_clean())),
+                );
+                Ok(state_object)
+            }
+            Ok(None) => {
+                // this state object is not exist in patriciaTrie, maybe you need to crate a new contract
+                Err(Error::AccountNotExist)
+            }
+            Err(_) => Err(Error::TrieError),
+        }
+    }
+
+    pub fn set_storage(&mut self, address: &Address, key: H256, value: H256) -> Result<(), Error> {
+        if self.storage_at(address, &key) != Some(value) {
+            let contain_key = self.cache.borrow().contains_key(address);
+            if !contain_key {
+                let trie = PatriciaTrie::from(&mut self.db, RLPNodeCodec::default(), &self.root.0)
+                    .or(Err(Error::TrieError))?;
+                match trie.get(&address) {
+                    Ok(Some(rlp)) => {
+                        let mut state_object = StateObject::from_rlp(&rlp)?;
+                        state_object.set_storage(key, value);
+                        self.insert_cache(address, StateObjectEntry::new_dirty(Some(state_object)));
+                    }
+                    Ok(None) => {
+                        // this state object is not exist in patriciaTrie, maybe you need to crate a new contract
+                        return Err(Error::AccountNotExist);
+                    }
+                    Err(_err) => {
+                        return Err(Error::TrieError);
+                    }
+                }
+            }
+        }
+        self.add_checkpoint(address);
+
+        if let Some(ref mut state_object_entry) = self.cache.borrow_mut().get_mut(address) {
+            match state_object_entry.state_object {
+                Some(ref mut state_object) => {
+                    state_object.set_storage(key, value);
+                    state_object_entry.status = ObjectStatus::Dirty;
+                }
+                None => panic!("state object always exist in cache."),
+            }
+        }
+        Ok(())
+    }
+
+    pub fn insert_cache(&self, address: &Address, state_object_entry: StateObjectEntry) {
+        let is_dirty = state_object_entry.is_dirty();
+        let old_entry = self
+            .cache
+            .borrow_mut()
+            .insert(*address, state_object_entry.clone_dirty());
+
+        if is_dirty {
+            if let Some(checkpoint) = self.checkpoints.borrow_mut().last_mut() {
+                checkpoint.entry(*address).or_insert(old_entry);
+            }
+        }
+    }
+
+    pub fn commit(&mut self) -> Result<(), Error> {
+        assert!(self.checkpoints.borrow().is_empty());
+
```

Comment:
If I commit twice in a row, I don't do anything in the middle, here panic?

Reply:
Layers for state:

```
| Cache       |
| CheckPointN |
| ...         |
| CheckPoint2 |
| CheckPoint1 |
| Disk DB     |
```

Commit means flush data from `Cache` to `Disk DB`, we must ensure that all checkpoints have been discarded or reverted(`self.checkpoints.borrow().is_empty()`).